src/Core/IRuleSettingsProvider.cs

```diff
@@ -0,0 +1,105 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.Integration;
+
+namespace SonarLint.VisualStudio.Core
+{
+    public interface IRuleSettingsProviderFactory
+    {
+        /// <summary>
+        /// Create <see cref="IRuleSettingsProvider"/> for a specific language
+        /// </summary>
+        /// <remarks>
+        /// In connected mode we need to know which file to load and we get it from <see cref="Language.FileSuffixAndExtension"/>.
+        /// </remarks>
+        IRuleSettingsProvider Get(Language language);
+    }
+
+    public interface IRuleSettingsProvider
+    {
+        /// <summary>
+        /// Get rule settings specified in connected mode, or in the settings.json file
+        /// </summary>
+        RulesSettings Get();
+    }
+
+    public class RuleSettingsProvider : IRuleSettingsProvider
+    {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly IUserSettingsProvider userSettingsProvider;
+        private readonly IRulesSettingsSerializer serializer;
+        private readonly Language language;
+        private readonly ILogger logger;
+
+        public RuleSettingsProvider(IActiveSolutionBoundTracker activeSolutionBoundTracker,
+            IUserSettingsProvider userSettingsProvider,
+            IRulesSettingsSerializer serializer,
+            Language language,
+            ILogger logger)
+        {
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.userSettingsProvider = userSettingsProvider;
+            this.serializer = serializer;
+            this.language = language;
+            this.logger = logger;
+        }
+
+        public RulesSettings Get()
+        {
+            RulesSettings settings = null;
+
+            // If in connected mode, look for rule settings in the .sonarlint/sonarqube folder.
+            var binding = this.activeSolutionBoundTracker.CurrentConfiguration;
+
+            if (binding != null && binding.Mode != SonarLintMode.Standalone)
+            {
+                settings = FindConnectedModeSettings(language, binding);
+                if (settings == null)
+                {
+                    logger.WriteLine(CoreStrings.UnableToLoadConnectedModeSettings);
+                }
+                else
+                {
+                    logger.WriteLine(CoreStrings.UsingConnectedModeSettings);
+                }
+            }
+
+            // If we are not in connected mode or couldn't find the connected mode settings then fall back on the standalone settings.
+            settings = settings ?? userSettingsProvider.UserSettings.RulesSettings;
+
+            return settings;
+        }
+
+        private RulesSettings FindConnectedModeSettings(Language language, BindingConfiguration binding)
+        {
+            if (language == null)
```

Comment:
I will re-add the debug assert that was in the original code, accidentally removed